rails i18n　プラクティス

jepe-w opened the pull request.

jepe-w committed:
bookアプリを日本語対応

i18nを使って英語/日本語の対応

jepe-w committed:
文言修正。一部未対応箇所の改善

pushができなかったため一から作り直したことで、index画面の文言が対応していなかった点と、「新規登録」の文言がindex画面とnew画面でゆれていたので統一

maedana commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,35 @@
+      book: "書籍"
+    attributes:
+      book:
+        title: "タイトル"
```
これがあるにも関わらず、`books.index.title`があったりしますね。DBのカラムにあたるものはこの翻訳を使うようにしましょう。他にも複数ありますね。課題の

> [Rails 国際化 (i18n) API - Railsガイド](https://railsguides.jp/i18n.html) を読み、使えそうなメソッドがあれば積極的に使う

> 翻訳を追加したり編集したりする際は、アプリケーション全体で共通化する翻訳か、特定のモデルに特化した翻訳か、どちらにするのかを意識しましょう。
アプリ全体で共通化する翻訳なら、翻訳に特定のモデル名が含まれないようにする
特定のモデルに特化する翻訳なら、キーにモデル名が含まれるようにする

といったあたりを意識する必要があります。

maedana commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,35 @@
+  books:
+    index:
+      Books: "書籍一覧"
+      title: "タイトル :"
```
上述した対応をすればそもそも不要なキーになりますが、仮に必要であったとしても`:`のような部分はhtml側に書けばよいかと。

maedana commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,35 @@
+      memo: "メモ :"
+      author: "著者名 :"
+      picture: "画像 :"
+      Show this book: "この本を表示"
```
キー名はスネークケースで命名しましょう。あとキー名に動詞はあまり使わないかなというのと、具体的なキー名にしすぎると、`この本を表示`の内容が変わったときにキー名と割りててる翻訳で整合性が取れなくなったりします。具体的すぎずもっと抽象的な命名がよいかと思います。

jepe-w committed:
重複した翻訳を削除。キー名を改善

maedana commented on app/views/books/_book.html.erb:
```diff
@@ -1,21 +1,21 @@
 <div id="<%= dom_id book %>">
   <p>
-    <strong>Title:</strong>
+    <strong><%= t 'activerecord.attributes.book.title' %> :</strong>
```
キー名としては間違ってないのですが、前回書いた
> Rails 国際化 (i18n) API - Railsガイド を読み、使えそうなメソッドがあれば積極的に使う

を対応したいところです。さらに具体的にいうと
https://railsguides.jp/i18n.html#active-recordモデルで翻訳を行なう
のあたりですね。よーく読んでみましょう。

maedana commented on config/locales/views/ja.yml:
```diff
@@ -0,0 +1,8 @@
+            new: "新規登録"
+            edit: "編集"
+            to_list: "一覧に戻る"
+            delete: "削除"
```
なんだか妙にindent深いですね。半角スペース2文字にしたいですね

jepe-w committed:
views/ja.ymlの作り直し

jepe-w committed:
Model.human_attribute_name(attribute)の活用

maedana commented on app/views/books/_book.html.erb:
```diff
@@ -1,21 +1,21 @@
 <div id="<%= dom_id book %>">
   <p>
-    <strong>Title:</strong>
+    <strong><%= t Book.human_attribute_name('title') %> :</strong>
```
これだとtメソッドの引数に「タイトル」みたいなものを渡してることになっちゃいます

jepe-w committed:
指摘箇所の修正

<strong><%= t Book.human_attribute_name('title') %> :</strong> ======> <strong><%= Book.human_attribute_name('title') %> :</strong> (tメソッドの削除)
